In [23]:
from google.colab import drive
drive.mount('/content/drive')


Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
# Task 1

In [24]:
import nltk
nltk.download('popular')


[nltk_data] Downloading collection 'popular'
[nltk_data]    | 
[nltk_data]    | Downloading package cmudict to /root/nltk_data...
[nltk_data]    |   Package cmudict is already up-to-date!
[nltk_data]    | Downloading package gazetteers to /root/nltk_data...
[nltk_data]    |   Package gazetteers is already up-to-date!
[nltk_data]    | Downloading package genesis to /root/nltk_data...
[nltk_data]    |   Package genesis is already up-to-date!
[nltk_data]    | Downloading package gutenberg to /root/nltk_data...
[nltk_data]    |   Package gutenberg is already up-to-date!
[nltk_data]    | Downloading package inaugural to /root/nltk_data...
[nltk_data]    |   Package inaugural is already up-to-date!
[nltk_data]    | Downloading package movie_reviews to
[nltk_data]    |     /root/nltk_data...
[nltk_data]    |   Package movie_reviews is already up-to-date!
[nltk_data]    | Downloading package names to /root/nltk_data...
[nltk_data]    |   Package names is already up-to-date!
[nltk_data]    | Do

True

In [25]:
nltk.download('book')
from nltk.book import *

[nltk_data] Downloading collection 'book'
[nltk_data]    | 
[nltk_data]    | Downloading package abc to /root/nltk_data...
[nltk_data]    |   Unzipping corpora/abc.zip.
[nltk_data]    | Downloading package brown to /root/nltk_data...
[nltk_data]    |   Unzipping corpora/brown.zip.
[nltk_data]    | Downloading package chat80 to /root/nltk_data...
[nltk_data]    |   Unzipping corpora/chat80.zip.
[nltk_data]    | Downloading package cmudict to /root/nltk_data...
[nltk_data]    |   Package cmudict is already up-to-date!
[nltk_data]    | Downloading package conll2000 to /root/nltk_data...
[nltk_data]    |   Unzipping corpora/conll2000.zip.
[nltk_data]    | Downloading package conll2002 to /root/nltk_data...
[nltk_data]    |   Unzipping corpora/conll2002.zip.
[nltk_data]    | Downloading package dependency_treebank to
[nltk_data]    |     /root/nltk_data...
[nltk_data]    |   Unzipping corpora/dependency_treebank.zip.
[nltk_data]    | Downloading package genesis to /root/nltk_data...
[nltk_d

*** Introductory Examples for the NLTK Book ***
Loading text1, ..., text9 and sent1, ..., sent9
Type the name of the text or sentence to view it.
Type: 'texts()' or 'sents()' to list the materials.
text1: Moby Dick by Herman Melville 1851
text2: Sense and Sensibility by Jane Austen 1811
text3: The Book of Genesis
text4: Inaugural Address Corpus
text5: Chat Corpus
text6: Monty Python and the Holy Grail
text7: Wall Street Journal
text8: Personals Corpus
text9: The Man Who Was Thursday by G . K . Chesterton 1908


In [ ]:
#first few words of Moby Dick
print("Sample text from Moby Dick:")
print(text1[:100])

Sample text from Moby Dick:
['[', 'Moby', 'Dick', 'by', 'Herman', 'Melville', '1851', ']', 'ETYMOLOGY', '.', '(', 'Supplied', 'by', 'a', 'Late', 'Consumptive', 'Usher', 'to', 'a', 'Grammar', 'School', ')', 'The', 'pale', 'Usher', '--', 'threadbare', 'in', 'coat', ',', 'heart', ',', 'body', ',', 'and', 'brain', ';', 'I', 'see', 'him', 'now', '.', 'He', 'was', 'ever', 'dusting', 'his', 'old', 'lexicons', 'and', 'grammars', ',', 'with', 'a', 'queer', 'handkerchief', ',', 'mockingly', 'embellished', 'with', 'all', 'the', 'gay', 'flags', 'of', 'all', 'the', 'known', 'nations', 'of', 'the', 'world', '.', 'He', 'loved', 'to', 'dust', 'his', 'old', 'grammars', ';', 'it', 'somehow', 'mildly', 'reminded', 'him', 'of', 'his', 'mortality', '.', '"', 'While', 'you', 'take', 'in', 'hand', 'to', 'school', 'others', ',']


In [26]:
from nltk.probability import FreqDist

# 1(a) Vocabulary Size
vocabulary_size = len(set(text1))
print("Vocabulary Size:", vocabulary_size)

# or unique words
word_freq = FreqDist(text1)
print("Number of unique words (including punctuation):", len(word_freq))



Vocabulary Size: 19317
Number of unique words (including punctuation): 19317


In [27]:
# (b) Total Words
total_words = len(text1)
print("Total Words:", total_words)

Total Words: 260819


In [28]:
# (c) the number of times a word is present in the novella
word_freq = FreqDist(text1)
word_count_dict = dict(word_freq)

with open("/content/drive/MyDrive/440/word_frequencies.txt", "w") as f:
    for word, count in word_count_dict.items():
        f.write(f"{word}: {count}\n")

print("Word frequencies saved to 'word_frequencies.txt'")


Word frequencies saved to 'word_frequencies.txt'


link to the **word_frequencies.txt** file: https://drive.google.com/file/d/1ualNSoEfzcLWiLUylNzPIwC1NWDC8d_K/view?usp=drive_link

In [29]:
# (d) Top 10 Most Frequent Words (Including Punctuation)
top_10_words = word_freq.most_common(10)
print("Top 10 Most Frequent Words (including punctuation):")
print(top_10_words)


Top 10 Most Frequent Words (including punctuation):
[(',', 18713), ('the', 13721), ('.', 6862), ('of', 6536), ('and', 6024), ('a', 4569), ('to', 4542), (';', 4072), ('in', 3916), ('that', 2982)]


In [30]:
# (e) Top 10 Most Frequent Words (Excluding Punctuation)
words_only = [word for word in text1 if word.isalpha()]
word_freq_no_punct = FreqDist(words_only)

top_10_no_punct = word_freq_no_punct.most_common(10)
print("Top 10 Most Frequent Words (excluding punctuation):")
print(top_10_no_punct)

Top 10 Most Frequent Words (excluding punctuation):
[('the', 13721), ('of', 6536), ('and', 6024), ('a', 4569), ('to', 4542), ('in', 3916), ('that', 2982), ('his', 2459), ('it', 2209), ('I', 2124)]


In [ ]:
# Task 2

In [40]:
# (a)
import pandas as pd
from nltk.tokenize import word_tokenize, sent_tokenize

nltk.download('punkt')
nltk.download('punkt_tab')

nltk.data.path.append('/root/nltk_data')

file_path = "/content/drive/MyDrive/440/IMDB Dataset.csv"
df = pd.read_csv(file_path)

# total number of reviews
total_reviews = len(df)
print(f"Total number of reviews: {total_reviews}")

df['word_count'] = df['review'].apply(lambda x: len(word_tokenize(x)) if isinstance(x, str) else 0)
df['sentence_count'] = df['review'].apply(lambda x: len(sent_tokenize(x)) if isinstance(x, str) else 0)

longest_review_words_index = df['word_count'].idxmax()
longest_review_sentences_index = df['sentence_count'].idxmax()

longest_review_words = df.loc[longest_review_words_index]
longest_review_sentences = df.loc[longest_review_sentences_index]

print("\nLongest review in terms of words:")
print(f"Row: {longest_review_words_index}")
print(longest_review_words)

print("\nLongest review in terms of sentences:")
print(f"Row: {longest_review_sentences_index}")
print(longest_review_sentences)


[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package punkt_tab to /root/nltk_data...
[nltk_data]   Package punkt_tab is already up-to-date!


Total number of reviews: 50000

Longest review in terms of words:
Row: 40521
review            There's a sign on The Lost Highway that says:<...
sentiment                                                  positive
word_count                                                     2911
sentence_count                                                   65
Name: 40521, dtype: object

Longest review in terms of sentences:
Row: 39182
review            Smallville episode Justice is the best episode...
sentiment                                                  positive
word_count                                                      298
sentence_count                                                  282
Name: 39182, dtype: object


In [43]:
 # b

import string
from nltk.corpus import stopwords
import re

nltk.download('stopwords')

stop_words = set(stopwords.words('english'))

def clean_review(text):
    if isinstance(text, str):
        text_no_punctuation = text.translate(str.maketrans('', '', string.punctuation))
        word_tokens = text_no_punctuation.split()
        filtered_tokens = [word for word in word_tokens if word.lower() not in stop_words]
        return " ".join(filtered_tokens)
    else:
        return ""

df['cleaned_review'] = df['review'].apply(clean_review)


[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


In [44]:
# c

if 'cleaned_review' in df.columns:
    df['cleaned_review'] = df['cleaned_review'].apply(lambda x: x.lower() if isinstance(x, str) else "")

    output_text_file = '/content/drive/MyDrive/440/cleaned_reviews.txt'
    with open(output_text_file, 'w', encoding='utf-8') as f:
        for review in df['cleaned_review']:
            f.write(review.strip() + '\n')  # Writing each review as a new line in the text file


#     --------------
    with open(output_text_file, 'r', encoding='utf-8') as f:
        lines = f.readlines()

    if len(lines) == len(df):
        print(f"All {len(lines)} reviews successfully saved to {output_text_file}.")
    else:
        print(f"Some reviews might be missing. Saved {len(lines)} out of {len(df)} reviews.")
else:
    print("Column 'cleaned_review' not found in the dataset.")


All 50000 reviews successfully saved to /content/drive/MyDrive/440/cleaned_reviews.txt.


link to the **cleaned_reviews.txt** file: https://drive.google.com/file/d/135yXHFZx6YqsEZRZA1IAucJzfuYu-umB/view?usp=drive_link

In [ ]:
# Task 3

In [46]:
# a
from sklearn.feature_extraction.text import TfidfVectorizer

vectorizer = TfidfVectorizer()
tfidf_matrix = vectorizer.fit_transform(df['cleaned_review'])



In [47]:
feature_names_file = "/content/drive/MyDrive/440/tfidf_feature_names.txt"
matrix_file = "/content/drive/MyDrive/440/tfidf_matrix_sparse.txt"

with open(feature_names_file, 'w', encoding='utf-8') as file:
    file.write("Feature Names:\n")
    file.write("\n".join(vectorizer.get_feature_names_out()))


# sparse matrix
with open(matrix_file, 'w', encoding='utf-8') as file:
    file.write("TF-IDF Sparse Matrix:\n")
    for i, j in zip(*tfidf_matrix.nonzero()):  # nonzero() for sparse indices
        file.write(f"({i}, {j}) {tfidf_matrix[i, j]}\n")


print("TF-IDF vectorization complete using sparse operations.")
print(f"Feature names saved to {feature_names_file}.")
print(f"Sparse TF-IDF Matrix saved to {matrix_file}.")


TF-IDF vectorization complete using sparse operations.
Feature names saved to /content/drive/MyDrive/440/tfidf_feature_names.txt.
Sparse TF-IDF Matrix saved to /content/drive/MyDrive/440/tfidf_matrix_sparse.txt.


link to [**tfidf_feature_names.txt**](https://drive.google.com/file/d/1josOj4qcNWkfjRBZSYChD09GwAFjyvzS/view?usp=drive_link) file   and   
link to [**tfidf_feature_names.txt**](https://drive.google.com/file/d/1yC0VHtj5gxeI9ihw3qZj0p4qtCH3IYnO/view?usp=drive_link) file

In [48]:
matrix_shape = tfidf_matrix.shape

print(f"The shape of the TF-IDF matrix is: {matrix_shape}")
print(f"Number of documents: {matrix_shape[0]}")
print(f"Number of features (vocabulary size): {matrix_shape[1]}")


The shape of the TF-IDF matrix is: (50000, 180355)
Number of documents: 50000
Number of features (vocabulary size): 180355


In [49]:

# b

import numpy as np

def load_glove_embeddings(file_path):
    embeddings = {}
    with open(file_path, 'r', encoding='utf-8') as f:
        for line in f:
            values = line.split()
            word = values[0]
            vector = np.array(values[1:], dtype='float32')
            embeddings[word] = vector
    return embeddings


glove_file = "/content/drive/MyDrive/440/glove.6B.100d.txt"
glove_embeddings = load_glove_embeddings(glove_file)

print(f"Total words loaded: {len(glove_embeddings)}")



Total words loaded: 400000


In [50]:
output_file = "/content/drive/MyDrive/440/glove_vectors.txt"

with open(output_file, 'w', encoding='utf-8') as f:
    for word, vector in glove_embeddings.items():
        vector_string = " ".join(map(str, vector))
        f.write(f"{word} {vector_string}\n")

print(f"All word vectors have been saved to {output_file}.")

All word vectors have been saved to /content/drive/MyDrive/440/glove_vectors.txt.


link to the **glove_vectors.txt** file: https://drive.google.com/file/d/1wWokLW48dzOlXuoRGKX2NWOV-banLQpr/view?usp=drive_link

In [51]:
words = ["man", "woman", "cat", "dog", "king", "queen"]
vectors = {word: glove_embeddings[word] for word in words}

def cosine_similarity(vec1, vec2):
    dot_product = np.dot(vec1, vec2)
    norm_vec1 = np.linalg.norm(vec1)
    norm_vec2 = np.linalg.norm(vec2)
    return dot_product / (norm_vec1 * norm_vec2)

similarities = {
    "man_woman": cosine_similarity(vectors["man"], vectors["woman"]),
    "cat_dog": cosine_similarity(vectors["cat"], vectors["dog"]),
    "king_queen": cosine_similarity(vectors["king"], vectors["queen"]),
}

king_minus_man_plus_woman = vectors["king"] - vectors["man"] + vectors["woman"]

similarity_to_queen = cosine_similarity(king_minus_man_plus_woman, vectors["queen"])

print("Cosine Similarities:")
for pair, similarity in similarities.items():
    print(f"{pair}: {similarity:.4f}")

# print("Queen:", vectors["queen"])
# print("King - Man + Woman:", king_minus_man_plus_woman)
print(f"\nSimilarity between King - Man + Woman and Queen: {similarity_to_queen:.4f}")

Cosine Similarities:
man_woman: 0.8323
cat_dog: 0.8798
king_queen: 0.7508

Similarity between King - Man + Woman and Queen: 0.7834


Yes, they reflect what we learnt in class as a cosine similarity close to 1 (0.7834) indicates that the result vector is very similar to the vector for "Queen."